osf/models/mixins.py

```diff
@@ -332,12 +332,28 @@ def add_node_link(self, node, auth, save=True):
         :param bool save: Save changes
         :return: Created pointer
         """
-        # Fail if node already in nodes / pointers. Note: cast node and node
-        # to primary keys to test for conflicts with both nodes and pointers
-        # contained in `self.nodes`.
-        if NodeRelation.objects.filter(parent=self, child=node, is_node_link=True).exists():
+        if (self._id == node._id):
             raise ValueError(
-                'Link to node {0} already exists'.format(node._id)
+                'Cannot link node \'{}\' to itself.'.format(self._id)
+            )
+        existant_relation = NodeRelation.objects.filter(parent=self, child=node).first()
+        if existant_relation and existant_relation.is_node_link:
+            raise ValueError(
+                'Target Node \'{}\' already pointed to by \'{}\'.'.format(node._id, self._id)
+            )
+        elif existant_relation and not existant_relation.is_node_link:
+            raise ValueError(
+                'Target Node \'{}\' is already a child of \'{}\'.'.format(node._id, self._id)
+            )
+
+        existant_relation = NodeRelation.objects.filter(parent=node, child=self).first()
+        if existant_relation and existant_relation.is_node_link:
+            raise ValueError(
+                'Node \'{}\' is linked to target node \'{}\'.'.format(node._id, self._id)
+            )
+        elif existant_relation and not existant_relation.is_node_link:
+            raise ValueError(
+                'Node \'{}\' is already a child of target node \'{}\'.'.format(node._id, self._id)
```

Comment:
This could probably be more DRY! you could consolidate into a method that takes in which node is the child, and which node is the parent, or something like that, and then just call it twice.  This would help make sure your error messages are more similar - the errors where they're both node links should have parallel structure.